com.unity.ml-agents.extensions/Runtime/Input/InputActuatorComponent.cs

```diff
@@ -0,0 +1,376 @@
+
+            // Load the device layout based on the controls we created previously.
+            if (string.IsNullOrEmpty(existingLayout))
+            {
+                InputSystem.RegisterLayoutMatcher(layoutName, new InputDeviceMatcher()
+                    .WithInterface($"^({interfaceName}[0-9]*)"));
+            }
+
+            // Actually create the device instance.
+            var device = InputSystem.AddDevice(
+                new InputDeviceDescription
+                {
+                    interfaceName = interfaceName,
+                    serial = $"{localId}"
+                }
+            );
+            return device;
+        }
+
+        /// <summary>
+        ///
+        /// </summary>
+        /// <param name="defaultMap"></param>
+        /// <param name="layoutName"></param>
+        /// <param name="localId"></param>
+        /// <param name="behaviorParameters"></param>
+        /// <returns></returns>
+        internal static IActuator[] GenerateActionActuatorsFromAsset(
+            InputActionMap defaultMap,
+            string layoutName,
+            uint localId,
+            BehaviorParameters behaviorParameters)
+        {
+            // TODO does this need to change based on the action map we use?
+            var builder = new InputControlLayout.Builder()
+                .WithName(layoutName)
+                .WithFormat(mlAgentsLayoutFormat);
+
+            var actuators = new List<IActuator>();
+
+            foreach (var action in defaultMap)
+            {
+                var actionLayout = InputSystem.LoadLayout(action.expectedControlType);
+
+                builder.AddControl(action.name)
+                    .WithLayout(action.expectedControlType)
+                    .WithFormat(actionLayout.stateFormat)
+                    .WithUsages($"MLAgentPlayer-{localId}");
+
+                var devicePath = InputControlPath.Separator + layoutName;
+
+                // Reasonably, the input system starts adding numbers after the first none numbered name
+                // is added.  So for device ID of 0, we use the empty string in the path.
+                var deviceId = localId == 0 ? string.Empty : "" + localId;
+                var path = $"{devicePath}{deviceId}{InputControlPath.Separator}{action.name}";
+                action.AddBinding(path,
+                    action.interactions,
+                    action.processors,
+                    mlAgentsControlSchemeName + localId);
+
+                var adaptor = (IRLActionInputAdaptor)Activator.CreateInstance(
+                    controlTypeToAdaptorType[actionLayout.type]);
+                actuators.Add(new InputActionActuator(
+                    behaviorParameters,
+                    action,
+                    adaptor));
+            }
+            var layout = builder.Build();
+            if (InputSystem.LoadLayout(layout.name) == null)
+            {
+                InputSystem.RegisterLayout(layout.ToJson(), layout.name);
```

Comment:
This looks peculiar. Layout builders are generally meant to just be registered as delegates.

```
InputSystem.RegisterLayoutBuilder(name: "MyLayout",
    buildMethod: () =>
    {
        var builder = new InputControlLayout.Builder();
        //....
        return builder.Build();
    };
```

Turning the layout into JSON and then registering that is certainly doable but the above method avoids some needless layout construction/loading.